tests/virt_autotest/virt_autotest_base.pm

```diff
@@ -85,29 +85,88 @@ sub generateXML {
     generateXML_from_data($data, \%xmldata);
 }
 
+sub save_test_configuration {
+    my ($self, $assert_pattern, $add_junit_log_flag, $upload_virt_log_flag, $log_dir, $compressed_log_name, $upload_guest_assets_flag) = @_;
+
+    $self->{assert_pattern}           = $assert_pattern;
+    $self->{add_junit_log_flag}       = $add_junit_log_flag;
+    $self->{upload_virt_log_flag}     = $upload_virt_log_flag;
+    $self->{log_dir}                  = $log_dir;
+    $self->{compressed_log_name}      = $compressed_log_name;
+    $self->{upload_guest_assets_flag} = $upload_guest_assets_flag;
+}
+
+#This is the subroutine called inside post_execute_script_run. It aims to do configurations have to be done after script
+#execution but before test assertion, for example, modifying boot options, communicating to peer with lock/unlcok and many other
+#things as long as they should be done and need to be done right after test execution or the following test steps may malfunction
+#without them. post_execute_script_configuration should be overriden in individual test modules.
+sub post_execute_script_configuration {
+    diag("You need to override this function post_execute_script_config in your test module");
+}
+
+#This is the subroutine called inside post_execute_script_run. This is introduced to faciliate tests that do not have or not convenient
+#to use assert pattern,  can not use directly returned output from execute_script_run or needs more customized way to manipuluate results.
+#So individual test modules can also have more control over how their test results should be extracted out, intead of just relying on a
+#singel assert pattern. post_execute_script_assertion needs to be overriden in individual test modules.
+sub post_execute_script_assertion {
+    diag("You need to override this function post_execute_script_assertion in your test module");
+}
+
+#This subroutine incorporates operations needs to be done after finishing executing script. It is further dividied into three parts,
+#including post_execute_script_configuration, upload_virt_logs and do test assertion. Test assertion can be done in two ways by using
+#assert pattern or newly introduced post_execute_script_assertion subroutine.
```

Comment:
If it is intended usage, can  a warning be added here that if assert_pattern is not empty, post_execute_script_assertion will not be applied?

Reply:
Done

Reply:
I do not see it in code yet...

Reply:
The two new diag() calls in post_execute_script_run

Reply:
Sorry, my initial comment was to suggest adding a warning in function description -- just in this comment part.

Reply:
Also done